[onert] When input shape.dim[k]==1 && signature[k]==-1, run model with input shape.dim[k]==1

hyunsik-yoon opened the pull request:
With this commit, unknown dim's default value will be 1.

**Why**: Previously, unknown dim must be set by calling `nnfw_set_input_tensorinfo()`. ~However, TFLite interpreter can run a model with unknown dim input by setting the default value of unknown dim to `1`.~ In case of TensorFlow family, TFLite Converts set the value of unknown dim to `1` so that TFLite can run it.

~**Impact**: at compilation stage, setting unknown dim to 1 sometimes works but sometimes does not.~

~This works: `add ( placeholder([None, 2]), placeholder([1, 2])`~
~This does not work: `concat (placeholder([2, None]), placeholder([2, 3]), axis=0)`~

~The second case will fail during validation at compilation phase. However, the second case can work if we call `nnfw_input_tensorinfo()` after compilation. Therefore, we need some tweak with which onert should continue when op validation fails at compilation time.~

**Impact**: With this PR, model input will be **always static**. 

A test case that test this will be committed after this PR.

Signed-off-by: Hyun Sik Yoon <hyunsik.yoon.1024@gmail.com>

hyunsik-yoon commented on runtime/onert/core/include/ir/Shape.h:
```diff
@@ -94,8 +95,36 @@ struct Shape
+   *                  - at compilation time, shape [1, 1] and [2,2] does not match with concat.
+   *                    However, this should not generate an error
+   */
+  std::set<int32_t> _changeable_dims;
```
using `std::vector` makes code complicated to keep the same rank of `_dimension`. So `set` was used.

glistening commented:
@hyunsik-yoon 

I am not sure _"TFLite interpreter can run a model with unknown dim input by setting the default value of unknown dim to `1`."_  Maybe **`tflite_converter`** sets unknown dim to `1` while setting signature `-1` for tensorflow `None`. However, model in #202 already has explicit tensor shape `1`. This PR is not the problem I would like to solve in #2028. I would like to fix "TestInputUnknownDimInputConcatModelLoaded.*" errors that occurred by #2028 patch. If there are much to do for fixing them, I would like to fix them with you.

For giving default value in onert interpreter, I am not against. However, I would like to get other member's opinion.

hyunsik-yoon commented:
> I am not sure "TFLite interpreter can run a model with unknown dim input by setting the default value of unknown dim to 1". tflite_converter is likely to set unknown dim to 1 while setting signature -1 for tensorflow None. I am not sure.

I rephrased the sentence. Thanks.

> Model in #202 already has explicit tensor shape 1.

I am not sure the relation of this PR and #202. Could you explain a bit more?

> I would like to get fix for "TestInputUnknownDimInputConcatModelLoaded.*" errors that occurred by #2028 patch. 

#2028 patch is not complete. We have to fix `Compiler.cc` just like in this PR. 
Also "TestInputUnknownDimInputConcatModelLoaded.*" will fail if we treat unknown dim == 1. (the case 2 I mentioned at https://github.com/Samsung/ONE/pull/2028#issuecomment-640989176)

> If there are much to do for fixing them, I would like to fix them with you.

I hope that this PR fixes issue with the model you're testing. Could you please try to run with this PR?

> @hyunsik-yoon Could you split this PR into at least two? 1) introducing something for information of tensor signature, 2) concat shape inference?

Only with fix of `base_loader.h` and `compiler.cc`, unknown dim test with concat will fail as I mentioned.

If you want this PR to be spitted into several PRs, the other way around will be three PRs but I am not sure I have to do that.

PR 1) remove the test not to fail with PR 2.
PR 2) modify base_loader.h, compiler.cc, shape.h
PR 3) add the revised test and concat code

glistening commented:
> I am not sure the relation of this PR and #202. Could you explain a bit more?

Oops. Typo. It should be #2028. I've fixed. Thank you.

I suggested splitting this PR has several changes, I like some, I don't like some.

For fixing shape inference error about `Concat`, I am oaky. 
For supporting default value, I am neutral.
For introducing `changeable...`, I prefer to introduce tensor signature like tensorflow. tensor signature could be used for more attributes about tensor, not only `changeable`.

glistening commented:
I took a look into `TestInputUnknownDimInputConcatModelLoaded.*`. I downloaded `unknown_dim_input_concat.zip`. model looks strange.

![image](https://user-images.githubusercontent.com/12553213/84220528-f580b980-ab0d-11ea-9185-624092adae1d.png)

Thus, I've run the model with `tflite_run`. I've got segmentation fault as I expected.

```
 $ Product/x86_64-linux.debug/out/bin/tflite_run unknown_dim_input_concat/unknown_dim_input_concat.tflite 
/home/brian/z/ONE/externals/tensorflow/tensorflow/lite/kernels/concatenation.cc:75 t->dims->data[d] != t0->dims->data[d] (3 != 1)
Node number 0 (CONCATENATION) failed to prepare.

Segmentation fault (core dumped)
```

@hyunsik-yoon Where did you get this model? Should we run this model?

hyunsik-yoon commented:
@glistening ,
> @hyunsik-yoon Where did you get this model? Should we run this model?

The model is test model I created. To run the model, please call the following before calling `sess->prepare()`.

```c++
diff --git a/tests/tools/tflite_run/src/tflite_run.cc b/tests/tools/tflite_run/src/tflite_run.cc
index be15a268..a1081f95 100644
--- a/tests/tools/tflite_run/src/tflite_run.cc
+++ b/tests/tools/tflite_run/src/tflite_run.cc
@@ -145,6 +145,13 @@ int main(const int argc, char **argv)
     sess = std::make_shared<nnfw::tflite::InterpreterSession>(interpreter.get());
   }
 
+  {
+    TfLiteQuantizationParams quant;
+    interpreter->SetTensorParametersReadWrite(
+        0, kTfLiteFloat32, "Input1",
+        {2, 3}, quant);
+  }
+
   try
   {
     if (mp)
```

Result:
```
odroid@odroid:~/work/2/ONE$ ./Product/out/bin/tflite_run --tflite ./Product/out/unittest_standalone/nnfw_api_gtest_models/unknown_dim_input_concat/unknown_dim_input_concat/unknown_dim_input_concat.tflite input tensor indices = [0,1,]
... run 0 takes 0.007 ms
output tensor indices = [2(max:11),]
===================================
MODEL_LOAD takes 0.001 ms
PREPARE    takes 0 ms
EXECUTE    takes 
- Min:  0.007 ms
- Max:  0.007 ms
- Mean: 0.007 ms
- GeoMean: 0.007 ms
===================================
```

hyunsik-yoon commented:
@glistening ,

> For introducing changeable..., I prefer to introduce tensor signature like tensorflow. tensor signature could be used for more attributes about tensor, not only changeable.

I see. I will crate a PR for `shape.h` with `std::vector signature` (not `std::set`) 
and the rest in next PR. 
Please, confirm.

glistening commented:
@hyunsik-yoon 

> The model is test model I created. To run the model, please call the following before calling `sess->prepare()`.

My point is that it is not valid model. ( tensorflow lite cannot handle this model since it is invalid. )
I would like to remove this test or replace this model that could be run with tflite interpreter.

hyunsik-yoon commented:
@glistening 

> However, I am not sure that is is required and right behaviour.
it is valid API. https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/interpreter.h#L58

For usage, please find `label_image` example in Tensorflow Lite:
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/examples/label_image

This example reads mobilenet with input [1, 244, 244, 3]
and calls `SetTensorParametersReadWrite(...)` with test image of width 517, height 606. (https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/examples/label_image/testdata)

glistening commented:
@hyunsik-yoon I am aware of the use case. I am not saying the resizing input is not required.

I don't understand why the change of this PR title:

**"Support default value (==1) of unknown dim of model input"**

would solve the problem.

Both model #2028 and mobilenet you've mentioned in https://github.com/Samsung/ONE/pull/2059#issuecomment-641753789 don't have unspecified dim in its input. The change of this PR title should not change the behavior for both model. 

Also, I am not sure **_the change of this PR title_** is required and right behaviour.

hyunsik-yoon commented:
1. terms

- `unknown dim`:

in fbs file:

```
  // Encodes `shape` with unknown dimensions. Unknown dimensions are
  // represented with -1.
  shape_signature:[int]; // Optional.
```

2. TFlite repo has the following changes about a month ago:
- https://github.com/tensorflow/tensorflow/commit/89072571ca67f813d60c6ab1f6707eb4ace6be0d
- they introduces `strict` field.
  - if `true`, only unknown dim can be resized.
  - if `false`, all dims can be resized.
  - default is `false`

hyunsik-yoon commented:
@glistening ,
I tried to change the title. Let's have a further offline discussion tomorrow.

@all,
I will close this PR and split this PR into some. (https://github.com/Samsung/ONE/pull/2059#issuecomment-641711923)

glistening commented:
@hyunsik-yoon Thank you for changing title. Now, the title is perfectly what I want to solve. On the way home yesterday, I thought I should write an issue that describes what is the problem I would like to solve rather putting a PR in #2028.

hyunsik-yoon commented:
@glistening 
Can I ask how you want this PR to be?
Do you want me to continue this work by splitting this PR into several?
Or do you need to discuss more with the issue you'll create (rather than continuing the change in this PR)?
I am fine with both. Please let me know your opinion. Thanks.

hyunsik-yoon commented:
Modified.

1. Since our `nnfw_set_input_tensorinfo()` can be called to overwrite any input dim (regardless of input `shape_signature`), we don't use `shape_signature`. So I removed it.

2. Concat test was modified to call `nnfw_set_input_tensorinfo()` before `nnfw_prepare()`. With the change in this PR, `nnfw_prepare()` will treat the first tensor as `[1, 1]` (all dims are unknown) and test will fail.

hseok-oh commented:
@hyunsik-yoon Merge ? or No merge? (please check label)

hyunsik-yoon commented:
@hseok-oh Thanks for letting me know. Would you merge this PR? Thanks.
